### ISSUE TITLE
Pyright does not spot a problem with non-hashable arguments passed into functions decorated with `@functools.cache`

### ISSUE THREAD

**Issue by squeaky-pl** [microsoft/pyright#7144]
`mypy 1.8.0 (compiled: yes)`
`pyright 1.1.348`
`Python 3.10.11`

Not sure if this is a bug or a feature request, mypy behaves differently here (and maybe more correctly?)

**Describe the bug**
Pyright does not spot a problem with non-hashable arguments passed into functions decorated with `@functools.cache`

**Code or Screenshots**
Consider the following code

```python
import functools
from os import PathLike


@functools.cache
def load_file(path: PathLike) -> str:
    return ""


def do_work(path: PathLike) -> str:
    return load_file(path)
```

https://github.com/python/typeshed/blob/main/stdlib/functools.pyi#L56

mypy error:

```
example.py:11: error: Argument 1 to "__call__" of "_lru_cache_wrapper" has incompatible type "PathLike[Any]"; expected "Hashable"  [arg-type] 
``` 

pyright shows no error.

Separetely I think mypy could actually detect the error at decoration time maybe, but this does not concern pyright.

**Comment by erictraut** [comment]
I think pyright's behavior here is correct based on the type specification and its rules of type compatibility.

`PathLike` is a protocol. An object compatible with `PathLike` is also compatible with `object`, and `object` is compatible with the `Hashable` protocol.

Hashability is admittedly an unusual protocol because a base class can be Hashable while its subclasses may not be. Normally, when a base class conforms to a protocol, all of its subclasses must also conform to the same protocol. To do otherwise is a violation of the LSP.

If you would like to discuss this in the [typing forum](https://discuss.python.org/c/typing/32) — and potentially argue for an amendment or addition to the Python typing spec, you're welcome to do so. But by my reading of the spec, pyright's behavior is correct, so I'm going to close this.

**Comment by squeaky-pl** [comment]
Understood. I really appreciate the time you spent to explain this to me in depth. Thank you.
